### DISCUSSION TITLE
[QUESTION] Persistent volumes - storage quotas

### DISCUSSION THREAD

**Discussion by dberardo-com** [longhorn/longhorn#4723] [Q&A]
## Question

Is there a way to enforce "hard quotas" on persistent volumes provisioned using longhorn as storage class? In my use case i want to limit total namespace storage to, say, 100 GB per namespace. In order to do this i would like to add dynamic provisioning of PV using PVs but limit the total amount of available storage per namespace to 100GB. Is there a way to achieve this using longhorn?

At the moment if i set the capacitiy param in the PVC / PV spec, the volume is created but storage limits are not really enforce, so capacity becomes a useless parameter.

I see that there is some way to acheive this in rancher's project: https://docs.ranchermanager.rancher.io/v2.5/pages-for-subheaders/manage-project-resource-quotas but is this possible also without installing rancher?

## Environment

 - Kubernetes version: k3s 1.22

## Additional context

* i am using k3s/longhorn on a bare metal installation
* in general i am pretty confused with the notion of "capacity" in k8s PVs, because the documentation is extremely vague about it and it seems to me that this param does pretty much nothing. Is anyone aware of other storage classes that can achieve this (IMO "simple") requirement?

**Comment by derekbit** [comment]
You probably can refer to the [resource quota](https://kubernetes.io/docs/tasks/administer-cluster/limit-storage-consumption) for the per-namespace hard quota. However, longhorn can take volumes' snapshots, so overall space usage of voumes in a namespace can execeed the `requests.storage` you set.

Another idea is configuring your hosts' disks with filesystems' or lvm's quota feature.

**Comment by dberardo-com** [comment]
thanks for the answer. indeed i see that longhorn's volumes already enforce capacity limits defined in PVs, meaning that the container sees the volume being mounted with the actual capacity as defined in the PVC. 

this is good, i will have a look at the resource you provided i think it can fulfill my use cases' needs.

is it possible to predict in advance how much storage will longhorn require for snapshots? and how to limit this?

**Reply by derekbit** [reply]
> is it possible to predict in advance how much storage will longhorn require for snapshots? and how to limit this?

It's not easy, because the each snapshot is sparse. Every volume can have at most 250 snapshots, and the maximum snapshot size is the volume size.
If you use recurring job to take snapshots, you can set the snapshot retain number to limit the space usage comsumed by snapshots.

**Reply by c3y1huang** [reply]
Just adding on top of the comment.

Snapshot storage depends on the data difference between the previous snapshot and the current. [Here](https://github.com/longhorn/longhorn/issues/1534#issuecomment-651545069) is a short example of how the actual data is calculated, and [here](https://longhorn.io/docs/1.3.2/volumes-and-nodes/volume-size/#example) is a detailed example.

**Reply by dberardo-com** [reply]
thanks for the clarification. is it possible to disable snapshotting on specific volumes, or maybe make it disabled by default?

also, i have created a 10Mi volume, but i see it mounted inside the container as a 5,5 Mi file system ("df -h" command). i assume this is a normal overhead and has nothing to do with "bad configurations" from my side. right?

**Reply by c3y1huang** [reply]
> thanks for the clarification. is it possible to disable snapshotting on specific volumes, or maybe make it disabled by default?

You can set up a recurring job and add it to a group, then use that recurring job group for specific volumes. Ref https://longhorn.io/docs/1.3.2/snapshots-and-backups/scheduling-backups-and-snapshots/

> also, i have created a 10Mi volume, but i see it mounted inside the container as a 5,5 Mi file system ("df -h" command). i assume this is a normal overhead and has nothing to do with "bad configurations" from my side. right?

Sounds right, [here](https://github.com/longhorn/longhorn/issues/2176) is a similar question.

**Comment by dberardo-com** [comment]
do longhorn volumes support linux storage quotas on the filesystem?

i am using a single RWX longhorn volume and i would like to impose some storage quota on the folders at the file system level on a linux based container (debian image or so). is this something supported by longhorn or is it independent from the storageclass.

cheers
